src/librustc_mir/dataflow/mod.rs

```diff
@@ -235,21 +238,24 @@ impl<'a, 'tcx: 'a, BD> DataflowAnalysis<'a, 'tcx, BD> where BD: BitDenotation
 
 impl<'b, 'a: 'b, 'tcx: 'a, BD> PropagationContext<'b, 'a, 'tcx, BD> where BD: BitDenotation
 {
-    fn walk_cfg(&mut self, in_out: &mut IdxSet<BD::Idx>) {
+    fn walk_cfg(&mut self, in_out: &mut IdxSet<BD::Idx>, dirty_queue: &mut WorkQueue<mir::BasicBlock>) {
         let mir = self.builder.mir;
-        for (bb_idx, bb_data) in mir.basic_blocks().iter().enumerate() {
-            let builder = &mut self.builder;
-            {
-                let sets = builder.flow_state.sets.for_block(bb_idx);
-                debug_assert!(in_out.words().len() == sets.on_entry.words().len());
-                in_out.clone_from(sets.on_entry);
-                in_out.union(sets.gen_set);
-                in_out.subtract(sets.kill_set);
-            }
-            builder.propagate_bits_into_graph_successors_of(
-                in_out, &mut self.changed, (mir::BasicBlock::new(bb_idx), bb_data));
+        while let Some(bb) = dirty_queue.pop() {
+            let bb_data: &BasicBlockData<'tcx> = &mir.basic_blocks()[*bb];
+                let builder = &mut self.builder;
+                {
+                    let sets = builder.flow_state.sets.for_block(bb_idx);
+                    debug_assert!(in_out.words().len() == sets.on_entry.words().len());
+                    in_out.clone_from(sets.on_entry);
+                    in_out.union(sets.gen_set);
+                    in_out.subtract(sets.kill_set);
+                }
+                builder.propagate_bits_into_graph_successors_of(
+                    in_out, &mut self.changed, (mir::BasicBlock::new(bb_idx), bb_data), dirty_queue);
```

Comment:
I don't think we need the `changed` flag anymore -- the only bit that matters is if the dirty queue is empty.